kitsune/wiki/utils.py

```diff
@@ -103,56 +108,68 @@ def _active_contributors_id(from_date, to_date, locale, product):
     return set(list(editors) + list(reviewers))
 
 
-def get_featured_articles(product=None, locale=settings.WIKI_DEFAULT_LANGUAGE):
-    """Returns 4 random articles from the most visited.
+def get_featured_articles(product=None, topics=None, locale=settings.WIKI_DEFAULT_LANGUAGE):
+    """Returns up to 4 random articles per topic from the most visited.
 
-    If a product is passed, it returns 4 random highly visited articles.
+    Args:
+        product: Optional product to filter by
+        topics: Optional iterable of topics to filter by
+        locale: Locale to get articles for, defaults to WIKI_DEFAULT_LANGUAGE
     """
+    # Get base queryset with all needed relations in one hit
     visits = (
         WikiDocumentVisits.objects.filter(period=LAST_7_DAYS)
+        .select_related("document")
+        .prefetch_related(
+            "document__products",
+            "document__topics",
+            Prefetch(
+                "document__translations",
+                queryset=(
+                    Document.objects.visible(
+                        locale=locale,
+                        current_revision__is_approved=True,
+                        is_archived=False,
+                        is_template=False,
+                    )
+                    if locale != settings.WIKI_DEFAULT_LANGUAGE
+                    else None
+                ),
+            ),
+        )
         .filter(
             document__restrict_to_groups__isnull=True,
             document__locale=settings.WIKI_DEFAULT_LANGUAGE,
+            document__is_archived=False,
+            document__is_template=False,
         )
         .exclude(document__products__slug__in=settings.EXCLUDE_PRODUCT_SLUGS_FEATURED_ARTICLES)
-        .exclude(document__is_archived=True)
-        .exclude(document__is_template=True)
-        .order_by("-visits")
-        .select_related("document")
     )
 
+    # Add product and topics filters to the database query
     if product:
-        visits = visits.filter(document__products__in=[product.id])
+        visits = visits.filter(document__products=product)
+    if topics:
+        visits = visits.filter(document__topics__in=topics).distinct()
```

Comment:
Why is distinct needed here? I assume that we do want the same article to appear in different topics if it's the most visited

Reply:
My thought was that the document we need is singular, but the results we'd get here would have duplicates because a document can have multiple topics. The document itself, once retrieved, will still have it's connection to the topics - so nothing is lost, but we have a smaller result to deal with. 

I guess technically it could be dropped if you know a good reason - is distinct() expensive for the lack of value? Or am I wrong that document visits don't change regardless of topics/products?

Reply:
We are querying `WikiDocumentVisits` and not `Document`. This means that each visit record related to one document b/c of the FK. I believe we can drop this without having any duplicates or affecting the query in any way